jsonschema-core.md

```diff
@@ -1714,62 +1714,35 @@ against the schema defined by this keyword.
 
 #### Keywords for Applying Subschemas Conditionally {#conditional}
 
-Three of these keywords work together to implement conditional application of a
-subschema based on the outcome of another subschema. The fourth is a shortcut
-for a specific conditional case.
+These keywords contain subschemas which are conditionally applied based on the
```

Comment:
I suppose this is a perfect example to me of https://github.com/json-schema-org/json-schema-spec/pull/1445#discussion_r1343191641 (I didn't look at this PR first, promise :)

But to me, this makes things way *less* clear than what was here I think.

Just my opinion of course.

Reply:
Is this a comment on this line or the concept of the PR as a whole?

Reply:
PR as a whole, sorry (for being unclear).

Reply:
The purpose of this change is actually to make the interaction of `if` with `then` and `else` more consistent.

All other keyword interactions (except `contains`/`minContains`, which I'm planning a PR for) are _described_ using annotations as the communication mechanism.  Implementations are required to actually do it that way, but it serves as a good description.  That's what https://github.com/json-schema-org/json-schema-spec/pull/1445 is about.

As they're currently defined, `then` and `else` depend on the _evaluation result_ of `if`.  Depending on an evaluation result of a sibling keyword doesn't exist anywhere else in the spec, and there's not a defined communication mechanism to describe that behavior.  But there is such a mechanism using annotations.  So redefining `then` and `else` to depend on the annotation result of `if` (which _is_ the evaluation result) makes sense.

Reply:
To me -- again just me -- it is undoubtedly going to be clear to *current* readers of the spec what the intended behavior is, regardless of formality. It's the way any programmer really thinks about these things -- annotations are (I think incontrovertibly) not how anyone already knows if/then/else to work.

I do hear you on consistency -- my personal preference would be to go in the *other* direction and reduce the number of other keywords defined using annotations. Again though, it's a stylistic thing, I find things generally clearer when they're not mentioned, but it's obviously easier case by case to look at something and compare.

Reply:
> my personal preference would be to go in the other direction and reduce the number of other keywords defined using annotations.

I'm open to this, but we'd need to either:

- define that keywords can depend on the evaluation result of other keywords
- describe this interaction in some other way

Currently, only "presence, absence, and value" dependencies are defined (which means I need to actually add "annotation result" to that list in #1445).

If we decide to backtrack on the annotations-as-communication-mechanism thing, I'd want to be sure that all keywords can be described without them.  The difficult ones are `unevaluated*`.

I'm going to hold these PRs and start a discussion.